h2o-core/src/main/java/water/H2ONode.java

```diff
@@ -581,13 +581,20 @@ private boolean keepSending() {
               _outgoingMsgQ.put(bb); // put back and give someone else a chance to deliver
               break; // terminate
             }
-            while( bb != null ) {         // while have an BB to process
-              assert !bb.isDirect() : "Direct BBs already got recycled";
```

Comment:
this does not seem like the right thing to do, according to the assert this case should have been handled above. did some other modification break this assert?

Reply:
To me this means we cannot work with direct BBs here because at the point we get here, they are already returned to the object pool in AutoBuffer because this is an asynchronous process.

Reply:
I have actually tried to put Runtime Exception to be thrown if the `bb` is direct and it was thrown

Reply:
yep, that is right, a direct byte buffer can definitely find its way in - the behavior you are describing is permissible

What I am saying is that the message of the assert means something like: "we cannot work with direct BBs here because at the point we get here, they are already returned to the object pool in AutoBuffer because this is an asynchronous process"

Reply:
I see, makes sense